### ISSUE TITLE
Non-static methods can't be referenced as Callables from static methods

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#91403]
### Tested versions

4.3 dev6 broken
4.3 dev5 works

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 (NVIDIA; 31.0.15.4633) - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 Threads)

### Issue description

```GDScript
extends Node

func _ready() -> void:
	static_method()

static func static_method():
	print(not_static)

func not_static():
	pass
```
Running this script in dev5 prints `Node::not_static`. Running in dev6 results in `Invalid access to property or key 'not_static' on a base object of type 'Nil'.`. The callable is not accessible even if you name the script and do `ClassName.not_static`.

### Steps to reproduce

1. Run the code above in 4.3 dev5
2. Run the code above in 4.3 dev6

### Minimal reproduction project (MRP)

See above.

**Comment by akien-mga** [comment]
CC @dalexeev @vnen

**Comment by dalexeev** [comment]
I think the old behavior is also incorrect, non-static members should not be accessible in a static context, there should be an analyzer error. If you try to call `not_static()` you will get the error `Cannot call non-static function "not_static()" from static function "static_method()".`.

**Comment by KoBeWi** [comment]
I was using it to obtain method name, i.e. `Engine.get_main_loop().call_group(GROUP_NAME, not_static.get_name())`. The alternative is hard-coding `&"non_static"`, which is less safe.

btw I forgot to mention that the new error is runtime-only. It doesn't appear in the editor.

**Comment by akien-mga** [comment]
Looking at the changelog, I suppose this might have been caused by #90223 ?

**Comment by dalexeev** [comment]
> Looking at the changelog, I suppose this might have been caused by #90223 ?

Yes, most likely. But I think this is a valid assumption for the compiler, we should not allow static access to non-static members at the analyzer level.

**Comment by vnen** [comment]
Yeah, the old behavior was wrong. For a Callable you need an object and a method name, but in a static constant there's no object.

Maybe we can consider adding a `nameof ` or something for this case if it's needed (see https://github.com/godotengine/godot-proposals/issues/837). But I don't think reverting to old behavior is a good idea.

Not showing the error in-editor is another issue and should be fixed too.

**Comment by KoBeWi** [comment]
I didn't test after the fix, because I assumed it was working, but seems like instead of allowing it, the fix just made the error appear in the editor too. Was that intended?

**Comment by dalexeev** [comment]
@KoBeWi Yes, that's how it was intended. `Callable` is not the same as function/method pointer. You can't access a non-static member of a class in a static context. #91412 only fixed an analyzer bug. The previous behavior was unintentional and was removed with the optimization for static methods in #90223.
